## Part 1: Preprocessing

In [1]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers
from sklearn.preprocessing import OneHotEncoder, LabelEncoder

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   
3   33        No  Travel_Frequently  Research & Development                 3   
4   27        No      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   
3          4  Life Sciences                        4          56   
4          1        Medical                        1          40   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   
3               3  ...                  3                        3   
4               3  ...                  3                        4   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   
3                 0                 8                      3               3   
4                 1                 6                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   
3               8                   7                        3   
4               2                   2                        2   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  
3                     0  
4                     2  

[5 rows x 27 columns]

In [20]:
# Determine the number of unique values in each column.
attrition_df.nunique()

Age                         43
Attrition                    2
BusinessTravel               3
Department                   3
DistanceFromHome            29
Education                    5
EducationField               6
EnvironmentSatisfaction      4
HourlyRate                  71
JobInvolvement               4
JobLevel                     5
JobRole                      9
JobSatisfaction              4
MaritalStatus                3
NumCompaniesWorked          10
OverTime                     2
PercentSalaryHike           15
PerformanceRating            2
RelationshipSatisfaction     4
StockOptionLevel             4
TotalWorkingYears           40
TrainingTimesLastYear        7
WorkLifeBalance              4
YearsAtCompany              37
YearsInCurrentRole          19
YearsSinceLastPromotion     16
YearsWithCurrManager        18
dtype: int64

In [23]:
attrition_df['Attrition'].value_counts()

Attrition
No     1233
Yes     237
Name: count, dtype: int64

In [24]:
attrition_df['Department'].value_counts()

Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64

In [3]:
# Create y_df with the Attrition and Department columns
y_df = attrition_df[["Attrition","Department"]]
y_df

Attrition              Department
0          Yes                   Sales
1           No  Research & Development
2          Yes  Research & Development
3           No  Research & Development
4           No  Research & Development
...        ...                     ...
1465        No  Research & Development
1466        No  Research & Development
1467        No  Research & Development
1468        No                   Sales
1469        No  Research & Development

[1470 rows x 2 columns]

In [4]:
# Create a list of at least 10 column names to use as X data
X_columns = ["Age",
"DistanceFromHome",
"HourlyRate",
"NumCompaniesWorked",
"PercentSalaryHike",
"TotalWorkingYears",
"YearsAtCompany",
"YearsInCurrentRole",
"YearsSinceLastPromotion",
"YearsWithCurrManager"]


# Create X_df using your selected columns
X_df = attrition_df[X_columns]

# Show the data types for X_df
X_df.dtypes


Age                        int64
DistanceFromHome           int64
HourlyRate                 int64
NumCompaniesWorked         int64
PercentSalaryHike          int64
TotalWorkingYears          int64
YearsAtCompany             int64
YearsInCurrentRole         int64
YearsSinceLastPromotion    int64
YearsWithCurrManager       int64
dtype: object

In [5]:
# Convert your X data to numeric data types however you see fit
# Add new code cells as necessary
X_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 10 columns):
 #   Column                   Non-Null Count  Dtype
---  ------                   --------------  -----
 0   Age                      1470 non-null   int64
 1   DistanceFromHome         1470 non-null   int64
 2   HourlyRate               1470 non-null   int64
 3   NumCompaniesWorked       1470 non-null   int64
 4   PercentSalaryHike        1470 non-null   int64
 5   TotalWorkingYears        1470 non-null   int64
 6   YearsAtCompany           1470 non-null   int64
 7   YearsInCurrentRole       1470 non-null   int64
 8   YearsSinceLastPromotion  1470 non-null   int64
 9   YearsWithCurrManager     1470 non-null   int64
dtypes: int64(10)
memory usage: 115.0 KB


In [6]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_df, y_df)

In [7]:
# Create a StandardScaler
X_scaler = StandardScaler()

# Fit the StandardScaler to the training data
X_scaler.fit(X_train)

# Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


In [8]:
## Create a OneHotEncoder for the Department column
department_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore', drop=None)

# Fit the encoder to the training data
department_encoder.fit(np.array(y_train['Department']).reshape(-1, 1))

# Create two new variables by applying the encoder
# to the training and testing data
y_department_train = department_encoder.transform(np.array(y_train['Department']).reshape(-1, 1))
y_department_test = department_encoder.transform(np.array(y_test['Department']).reshape(-1, 1))


In [9]:

# Create a OneHotEncoder for the Attrition column
attrition_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore', drop=None)

# Fit the encoder to the training data
attrition_encoder.fit(np.array(y_train['Attrition']).reshape(-1, 1))

# Create two new variables by applying the encoder
# to the training and testing data
y_attrition_train = attrition_encoder.transform(np.array(y_train['Attrition']).reshape(-1, 1))
y_attrition_test = attrition_encoder.transform(np.array(y_test['Attrition']).reshape(-1, 1))


## Create, Compile, and Train the Model

In [10]:
# Find the number of columns in the X training data

# Create the input layer
input_layer = layers.Input(shape=(X_train.shape[1],), name='input_features')

# Create at least two shared layers
shared_layer1 = layers.Dense(64, activation='relu')(input_layer)
shared_layer2 = layers.Dense(128, activation='relu')(shared_layer1)


In [11]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the hidden layer
dept_hidden_layer = layers.Dense(32, activation='relu')(shared_layer2)

# Create the output layer
dept_output = layers.Dense(3, activation='softmax', name='dept_output')(dept_hidden_layer)


In [12]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the hidden layer
attr_hidden_layer = layers.Dense(32, activation='relu')(shared_layer2)

# Create the output layer
attr_output = layers.Dense(2, activation='sigmoid', name='attr_output')(attr_hidden_layer)

In [13]:
# Create the model
model = Model(inputs=input_layer, outputs=[
    dept_output,
    attr_output
])

# Compile the model
model.compile(optimizer='adam',
              loss={'dept_output': 'categorical_crossentropy',
                    'attr_output': 'binary_crossentropy'},
              metrics={'dept_output': 'accuracy',
                       'attr_output': 'accuracy'})

# Summarize the model
model.summary()

Model: "model"
__________________________________________________________________________________________________
 Layer (type)                Output Shape                 Param #   Connected to                  
 input_features (InputLayer  [(None, 10)]                 0         []                            
 )                                                                                                
                                                                                                  
 dense (Dense)               (None, 64)                   704       ['input_features[0][0]']      
                                                                                                  
 dense_1 (Dense)             (None, 128)                  8320      ['dense[0][0]']               
                                                                                                  
 dense_2 (Dense)             (None, 32)                   4128      ['dense_1[0][0]']         

In [14]:
# Train the model
model.fit(
    X_train,
    {
        'dept_output': y_department_train,
        'attr_output': y_attrition_train
    },
    epochs=10,  # You can adjust the number of epochs based on your needs
    batch_size=32  # You can adjust the batch size based on your available memory
#    validation_split=0.2  # You can specify the validation split if you have a separate validation set
)


Epoch 1/10
35/35 [==============================] - 7s 4ms/step - loss: 1.9241 - dept_output_loss: 1.2063 - attr_output_loss: 0.7178 - dept_output_accuracy: 0.5590 - attr_output_accuracy: 0.7958
Epoch 2/10
35/35 [==============================] - 0s 5ms/step - loss: 1.2731 - dept_output_loss: 0.8309 - attr_output_loss: 0.4421 - dept_output_accuracy: 0.6171 - attr_output_accuracy: 0.8385
Epoch 3/10
35/35 [==============================] - 0s 5ms/step - loss: 1.2801 - dept_output_loss: 0.8457 - attr_output_loss: 0.4344 - dept_output_accuracy: 0.5862 - attr_output_accuracy: 0.8385
Epoch 4/10
35/35 [==============================] - 0s 5ms/step - loss: 1.2265 - dept_output_loss: 0.8004 - attr_output_loss: 0.4260 - dept_output_accuracy: 0.6243 - attr_output_accuracy: 0.8385
Epoch 5/10
35/35 [==============================] - 0s 4ms/step - loss: 1.2334 - dept_output_loss: 0.8135 - attr_output_loss: 0.4199 - dept_output_accuracy: 0.6198 - attr_output_accuracy: 0.8385
Epoch 6/10
35/35 [=======

In [17]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_test, {'dept_output': y_department_test, 'attr_output': y_attrition_test})
test_results

12/12 [==============================] - 0s 2ms/step - loss: 1.2510 - dept_output_loss: 0.8333 - attr_output_loss: 0.4177 - dept_output_accuracy: 0.6495 - attr_output_accuracy: 0.8370


[1.2509870529174805,
 0.8333047032356262,
 0.417682409286499,
 0.6494565010070801,
 0.8369565010070801]

In [19]:
# Print the accuracy for both department and attrition
print(f"Department Accuracy: {test_results[3]}")
print(f"Attrition Accuracy: {test_results[4]}")

Department Accuracy: 0.6494565010070801
Attrition Accuracy: 0.8369565010070801


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1.  This was a good question, as it made me do some digging.  Off the top, I didn't remember this being discussed much.  What I found is that this data is considered unbalanced.  That is, the attrition "target" column has a 16 vs 84 percent split across the values of YES and NO respectively. It is considered unbalanced because the minority class (value=YES in this case), is less than 20% of the data.  Rule of thumb is % being less than 10-20% is considered unbalanced (in this case 16%).  The minority class for department is even worse.  The HR department represents only 4% of that column data. Therefore, to specifically answer the question, "accuracy" is NOT a good metric to use with this data.

References are in my readme for these quotes.....

"Accuracy is a helpful metric when you deal with balanced classes and care about the overall model “correctness” and not the ability to predict a specific class."

How do you know if the dataset is balanced or unbalanced?
Consider the threshold: There is no fixed threshold that defines when data becomes imbalanced. However, a common rule of thumb is that if the minority class represents less than 10–20% of the total samples, the data is often considered imbalanced. Domain knowledge: Consider the context and domain-specific expectations.Mar 28, 2024



2. I used the two activation functions of sigmoid and softmax for the attr_output and dept_output layers respectively.  I used sigmoid on attrition because it is a single binary value (YES or NO: 1 or 0), and this is what sigmoid is best used for.  Conversely, I used softmax on the department column, because softmax is best used on categorical values and department had three catagories.


3. One approach to improve the results would be to select different columns for X.  That is, select ones that may have more correlation to the target columns. I selected columns that had many unique characteristics, and this may have lead to the lower results.  
Another approach I would take,  would be to improve the inbalance of the department column by using SMOTE (and perhaps other methods too). Synthetic minority oversampling, or SMOTE, I do recall be discussed.  It is used to increase the sampling data of the minority class, to bring the dataset into a better balance.
Finally, after trying SMOTE, I'd check into using other sampling techniques to improve upon the imbalance of the column.  A couple discussed were RU (random-under sampling) and RO (random-oveer sampling).       